SystemInformer/proctree.c

```diff
@@ -1451,6 +1486,155 @@ static VOID PhpUpdateProcessNodePriorityBoost(
     }
 }
 
+// This method was added to be called from the sort logic for sorting this column
+// It needs to refresh all nodes's (PPH_PROCESS_NODE) service names (in the sorting function referred to generically as node1, node2)
+// This was the logic that existed in the switch case for case PHPRTLC_SERVICENAMES:
+// Then instead of 'node' and 'processItem' utilize 'ProcessNode' and 'ProcessNode->ProcessItem' to referr to the various available attributes
+static VOID PhpUpdateProcessServiceNames(
+    _Inout_ PPH_PROCESS_NODE ProcessNode
+)
+{
+    if (!(ProcessNode->ValidMask & PHPN_SERVICENAMES))		// Bitwise flag for calling this update method
+    {
+        if (!ProcessNode->SvcNames)
+        {
+            ProcessNode->SvcNames = PhCreateString(L"");
+        }
+
+        if ((ProcessNode->ProcessItem->ServiceList && ProcessNode->ProcessItem->ServiceList->Count != 0))
+        {
+            PH_STRING_BUILDER stringBuilder;
+            PhInitializeStringBuilder(&stringBuilder, 1000);
+
+            // Based on itemtips.c	 :	serviceList = PhCreateList(Process->ServiceList->Count);
+            //		I don't think ServiceListLock is needed, since we are not modyfing this, just simply reading it.
+            //		In itemtips.c its inserting items
+
+            //PhAcquireQueuedLockShared(&ProcessNode->ProcessItem->ServiceListLock);
+
+            for (ULONG i = 0; i < ProcessNode->ProcessItem->ServiceList->Count; i++)
+            {
+                PPH_SERVICE_ITEM serviceItem = ProcessNode->ProcessItem->ServiceList->Items[i];
+
+                PhAppendStringBuilder(&stringBuilder, &serviceItem->Name->sr);
+                PhAppendStringBuilder2(&stringBuilder, L" (");
+                PhAppendStringBuilder(&stringBuilder, &serviceItem->DisplayName->sr);
+                PhAppendStringBuilder2(&stringBuilder, L")");
+
+                if (i + 1 < ProcessNode->ProcessItem->ServiceList->Count)
+                {
+                    PhAppendStringBuilder2(&stringBuilder, L"  |  ");
```

Comment:
I'm also not convinced why this is necessary. A service process can't be associated with multiple services, can it? Maybe I'm missing some knowledge on windows internals here.

Reply:
I understand there are services which host a library. Maybe multiple libraries. But each of those processes will have a 1-1 relationship with a service, name, and display name. Under what conditions would you have a _process_ associated with _two_ services? Perhaps you can provide me some reading material, screen shots, etc. so I can educate myself. If this _is_ the case then the `PID` columns on the `Services` tab would need to display multiple `PIDs`, right?

Reply:
Sure np. But no to multiple PIDs. 1 service isn’t split across multiple processes. I mean the inverse (and former as you mention): 1 PID to many services. This would be to effectively take the tooltip and flatten out the NewLines into whatever char-separated list for quickly viewing in-line in the process tree without having to hover over each. The value perhaps is niche and only useful when viewing services side-by-side based on their relative order in the process tree. But perhaps the value isn’t really there given how many ways there are to view associated services in the process tree view… just seems like to me (and perhaps a handful of others haha)

Reply:
Okay I did some research. I understand what you're saying now. Thanks for helping me understand I was wrong in my assumption.

Reply:
> A service process can't be associated with multiple services, can it?

It can. You won't see this often on modern machines (with 3.5+ GB RAM) because starting from Windows 10 RS2, SCM [disables this feature](https://docs.microsoft.com/en-us/windows/application-management/svchost-service-refactoring) for the sake of better isolation. System Informer fully supports listing all services hosted in a process and attributing threads to a specific service based on the `SubProcessTag` field in TEB (see `PhGetThreadServiceTag`).

![image](https://user-images.githubusercontent.com/30962924/174198767-4d2e3f06-dd4f-403f-ad2b-3894ea682309.png)

Reply:
TIL, thanks for the explanation @diversenok.

Reply:
> Okay I did some research. I understand what you're saying now. Thanks for helping me understand I was wrong in my assumption.

Cool. It’s NP. There so much you do I  would even know where to begin haha!

Reply:
> > A service process can't be associated with multiple services, can it?
> 
> It can. You won't see this often on modern machines (with 3.5+ GB RAM) because starting from Windows 10 RS2, SCM [disables this feature](https://docs.microsoft.com/en-us/windows/application-management/svchost-service-refactoring) for the sake of better isolation. System Informer fully supports listing all services hosted in a process and attributing threads to a specific service based on the `SubProcessTag` field in TEB (see `PhGetThreadServiceTag`).
> 
> ![image](https://user-images.githubusercontent.com/30962924/174198767-4d2e3f06-dd4f-403f-ad2b-3894ea682309.png)

Heyyyy! Happy to see ya here. I should look into what you mentioned and if this could be implemented cleaner (just the services for now). Or anyone else welcome to whom has greater PH / SI knowledge